File: src/main/java/com/github/stars/service/GithubSearchService.java
package com.github.stars.service;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.beans.factory.annotation.Qualifier;
import org.springframework.http.*;
import org.springframework.stereotype.Service;
import org.springframework.web.client.HttpClientErrorException;
import org.springframework.web.client.RestTemplate;

import javax.annotation.Resource;
import java.util.*;

/**
 * GitHub 搜索与 Star/Unstar 服务
 * 调用 GitHub REST API v3 进行仓库搜索和 Star 操作
 */
@Service
public class GithubSearchService {

    private static final Logger log = LoggerFactory.getLogger(GithubSearchService.class);
    private static final String GITHUB_API_BASE = "https://api.github.com";
    private static final String GITHUB_API_VERSION = "application/vnd.github.v3+json";

    @Autowired
    @Qualifier("longTimeoutRestTemplate")
    private RestTemplate restTemplate;

    @Resource
    private ObjectMapper objectMapper;

    @Resource
    private SystemConfigService configService;

    /**
     * 搜索 GitHub 仓库
     *
     * @param keyword  搜索关键词
     * @param language 编程语言过滤（可选）
     * @param sort     排序方式（stars, forks, updated）
     * @param page     页码
     * @param perPage  每页数量
     * @return Map 包含 total（总数）和 repos（仓库列表）
     */
    public Map<String, Object> searchRepos(String keyword, String language, String sort, int page, int perPage) {
        Map<String, Object> result = new LinkedHashMap<>();

        // 构建查询字符串
        StringBuilder queryBuilder = new StringBuilder();
        if (keyword != null && !keyword.trim().isEmpty()) {
            queryBuilder.append(keyword.trim());
        }
        if (language != null && !language.trim().isEmpty()) {
            queryBuilder.append(" language:").append(language.trim());
        }
        String query = queryBuilder.toString().trim();
        if (query.isEmpty()) {
            query = "stars:>1";
        }

        String url = GITHUB_API_BASE + "/search/repositories"
                + "?q=" + encodeQueryParam(query)
                + "&sort=" + (sort != null && !sort.isEmpty() ? sort : "stars")
                + "&page=" + page
                + "&per_page=" + perPage;

        log.info("GitHub 搜索请求: {}", url);

        try {
            HttpHeaders headers = buildAuthHeaders();
            HttpEntity<String> entity = new HttpEntity<>(headers);
            ResponseEntity<String> response = restTemplate.exchange(url, HttpMethod.GET, entity, String.class);

            if (response.getStatusCode() == HttpStatus.OK && response.getBody() != null) {
                JsonNode root = objectMapper.readTree(response.getBody());
                int totalCount = root.get("total_count").asInt(0);
                JsonNode items = root.get("items");

                List<Map<String, Object>> repos = new ArrayList<>();
                if (items != null && items.isArray()) {
                    for (JsonNode item : items) {
                        Map<String, Object> repo = extractRepoFields(item);
                        repos.add(repo);
                    }
                }

                result.put("total", totalCount);
                result.put("repos", repos);
            } else {
                result.put("total", 0);
                result.put("repos", new ArrayList<>());
            }
        } catch (HttpClientErrorException e) {
            if (e.getStatusCode() == HttpStatus.FORBIDDEN) {
                log.error("GitHub API 访问受限 (403): {}", e.getMessage());
                throw new RuntimeException("GitHub API rate limited");
            }
            log.error("GitHub API 请求异常: {}", e.getMessage());
            throw new RuntimeException("GitHub API 调用失败: " + e.getMessage());
        } catch (Exception e) {
            log.error("GitHub 搜索异常", e);
            throw new RuntimeException("GitHub 搜索失败: " + e.getMessage());
        }

        return result;
    }

    /**
     * Star 一个仓库
     *
     * @param owner 仓库所有者
     * @param repo  仓库名
     * @return true 表示已 Star（或已处于 Star 状态）
     */
    public boolean starRepo(String owner, String repo) {
        String url = GITHUB_API_BASE + "/user/starred/" + owner + "/" + repo;
        log.info("Star 仓库: {}/{}", owner, repo);

        try {
            HttpHeaders headers = buildAuthHeaders();
            headers.setContentLength(0);
            HttpEntity<String> entity = new HttpEntity<>(headers);
            ResponseEntity<String> response = restTemplate.exchange(url, HttpMethod.PUT, entity, String.class);

            int status = response.getStatusCodeValue();
            // 204: 成功 Star，304: 已经 Star 过
            if (status == 204 || status == 304) {
                return true;
            }
            log.warn("Star 仓库返回非预期状态码: {}", status);
            return false;
        } catch (HttpClientErrorException e) {
            if (e.getStatusCode() == HttpStatus.NOT_MODIFIED) {
                // 304 表示已经 Star 过
                return true;
            }
            log.error("Star 仓库失败 [{}/{}]: {}", owner, repo, e.getMessage());
            return false;
        } catch (Exception e) {
            log.error("Star 仓库异常 [{}/{}]", owner, repo, e);
            return false;
        }
    }

    /**
     * 取消 Star 一个仓库
     *
     * @param owner 仓库所有者
     * @param repo  仓库名
     * @return true 表示取消成功
     */
    public boolean unstarRepo(String owner, String repo) {
        String url = GITHUB_API_BASE + "/user/starred/" + owner + "/" + repo;
        log.info("取消 Star 仓库: {}/{}", owner, repo);

        try {
            HttpHeaders headers = buildAuthHeaders();
            HttpEntity<String> entity = new HttpEntity<>(headers);
            ResponseEntity<String> response = restTemplate.exchange(url, HttpMethod.DELETE, entity, String.class);

            int status = response.getStatusCodeValue();
            if (status == 204) {
                return true;
            }
            log.warn("取消 Star 返回非预期状态码: {}", status);
            return false;
        } catch (HttpClientErrorException e) {
            log.error("取消 Star 失败 [{}/{}]: {}", owner, repo, e.getMessage());
            return false;
        } catch (Exception e) {
            log.error("取消 Star 异常 [{}/{}]", owner, repo, e);
            return false;
        }
    }

    /**
     * 检查是否已 Star 某个仓库
     *
     * @param owner 仓库所有者
     * @param repo  仓库名
     * @return true 表示已 Star
     */
    public boolean checkStarred(String owner, String repo) {
        String url = GITHUB_API_BASE + "/user/starred/" + owner + "/" + repo;
        log.debug("检查 Star 状态: {}/{}", owner, repo);

        try {
            HttpHeaders headers = buildAuthHeaders();
            HttpEntity<String> entity = new HttpEntity<>(headers);
            ResponseEntity<String> response = restTemplate.exchange(url, HttpMethod.GET, entity, String.class);

            int status = response.getStatusCodeValue();
            return status == 204;
        } catch (HttpClientErrorException e) {
            if (e.getStatusCode() == HttpStatus.NOT_FOUND) {
                return false;
            }
            log.error("检查 Star 状态失败 [{}/{}]: {}", owner, repo, e.getMessage());
            return false;
        } catch (Exception e) {
            log.error("检查 Star 异常 [{}/{}]", owner, repo, e);
            return false;
        }
    }

    /**
     * 构建带有 GitHub Token 认证的 HTTP 请求头
     */
    private HttpHeaders buildAuthHeaders() {
        HttpHeaders headers = new HttpHeaders();
        headers.set("Accept", GITHUB_API_VERSION);
        headers.set("User-Agent", "GithubStars-Search");

        String token = configService.getValue("github.token");
        if (token != null && !token.isEmpty()) {
            headers.set("Authorization", "Bearer " + token);
        }

        return headers;
    }

    /**
     * URL 编码查询参数（Java 8 兼容方式）
     */
    private String encodeQueryParam(String value) {
        try {
            return java.net.URLEncoder.encode(value, "UTF-8")
                    .replace("+", "%20");
        } catch (Exception e) {
            return value;
        }
    }

    /**
     * 从 GitHub API 返回的 repo JSON 节点中提取关键字段
     */
    private Map<String, Object> extractRepoFields(JsonNode item) {
        Map<String, Object> repo = new LinkedHashMap<>();
        repo.put("id", item.get("id").asLong());
        repo.put("full_name", safeText(item, "full_name"));
        repo.put("description", safeText(item, "description"));
        repo.put("language", safeText(item, "language"));
        repo.put("stargazers_count", item.get("stargazers_count").asInt(0));
        repo.put("forks_count", item.get("forks_count").asInt(0));
        repo.put("html_url", safeText(item, "html_url"));
        repo.put("pushed_at", safeText(item, "pushed_at"));
        repo.put("created_at", safeText(item, "created_at"));

        // owner 信息
        JsonNode ownerNode = item.get("owner");
        if (ownerNode != null && !ownerNode.isNull()) {
            repo.put("owner_login", safeText(ownerNode, "login"));
            repo.put("owner_avatar_url", safeText(ownerNode, "avatar_url"));
        } else {
            repo.put("owner_login", "");
            repo.put("owner_avatar_url", "");
        }

        // topics
        List<String> topics = new ArrayList<>();
        JsonNode topicsNode = item.get("topics");
        if (topicsNode != null && topicsNode.isArray()) {
            for (JsonNode t : topicsNode) {
                topics.add(t.asText());
            }
        }
        repo.put("topics", topics);

        return repo;
    }

    /**
     * 安全获取 JSON 文本字段（null 返回空字符串）
     */
    private String safeText(JsonNode node, String field) {
        JsonNode fieldNode = node.get(field);
        if (fieldNode == null || fieldNode.isNull()) {
            return "";
        }
        return fieldNode.asText();
    }
}

File: packages/backend/src/github/github-search.service.ts
import { Injectable, Logger } from '@nestjs/common';
import { ConfigService } from '../config/config.service';
import { GithubApiService } from './github-api.service';
import { GithubRepoService } from './github-repo.service';

const GITHUB_API = 'https://api.github.com';

/** 未入库仓库详情的内存缓存有效期（10 分钟），降低 GitHub API 限流压力 */
const REPO_DETAIL_CACHE_TTL = 10 * 60 * 1000;
/** 缓存条目上限，超出后整体清空重建 */
const REPO_DETAIL_CACHE_MAX = 200;

@Injectable()
export class GithubSearchService {
    private readonly logger = new Logger(GithubSearchService.name);

    /** 未入库仓库详情缓存：fullName → { data, expiresAt } */
    private readonly repoDetailCache = new Map<string, { data: Record<string, unknown>; expiresAt: number }>();
    constructor(
        private readonly config: ConfigService,
        private readonly githubApi: GithubApiService,
        private readonly repoService: GithubRepoService,
    ) {}

    /**
     * 获取任意 GitHub 仓库详情（统一仓库详情页数据源）
     *
     * 数据策略：
     * 1. 本地库已收录 → 返回 DB 完整数据（含翻译），并按需拉取 README，inLibrary=true
     * 2. 未收录 → 实时调用 GitHub API 获取元数据 + README，组装为与 DB 模型同构的
     *    对象（id=null、无翻译字段），inLibrary=false，结果缓存 10 分钟
     *
     * 未入库的仓库只读展示，不写入数据库，避免污染 Star 列表。
     *
     * @param owner 仓库所有者
     * @param repo 仓库名
     * @returns 仓库详情对象（GithubRepo 同构 + inLibrary 标记）
     */
    async getRepoDetail(owner: string, repo: string): Promise<Record<string, unknown>> {
        const fullName = `${owner}/${repo}`;
        this.logger.log(`获取仓库详情: ${fullName}`);

        // 1. 本地库命中 → DB 数据（含翻译），按需拉取 README
        const local = await this.repoService.findByFullName(fullName);
        if (local) {
            if (!local.readmeFetched) {
                this.logger.log(`详情页触发 README 按需拉取: ${fullName}`);
                // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment -- ensureReadmeFetched 预存 Promise<any> 返回值
                const updated = await this.repoService.ensureReadmeFetched(Number(local.id));
                return { ...((updated ?? local) as Record<string, unknown>), inLibrary: true };
            }
            return { ...local, inLibrary: true };
        }

        // 2. 内存缓存命中
        const cached = this.repoDetailCache.get(fullName);
        if (cached && cached.expiresAt > Date.now()) {
            this.logger.log(`仓库详情命中缓存: ${fullName}`);
            return cached.data;
        }

        // 3. GitHub API 实时获取；README 失败不阻塞详情返回
        const mapped = await this.githubApi.fetchRepoByFullName(fullName);
        const readmeResult = await this.githubApi.fetchReadmeFromGitHub(fullName).catch((e: unknown) => {
            const msg = e instanceof Error ? e.message : String(e);
            this.logger.error(`仓库详情 README 拉取失败: ${fullName}, ${msg}`);
            return null;
        });

        const detail: Record<string, unknown> = {
            id: null,
            ...mapped,
            descriptionCn: null,
            readmeCn: null,
            readmeOriginal: readmeResult?.content ?? null,
            readmeFetched: true,
            inLibrary: false,
        };

        if (this.repoDetailCache.size >= REPO_DETAIL_CACHE_MAX) {
            this.repoDetailCache.clear();
        }
        this.repoDetailCache.set(fullName, { data: detail, expiresAt: Date.now() + REPO_DETAIL_CACHE_TTL });
        return detail;
    }

    /**
     * 构建 GitHub API 请求头
     *
     * 包含标准的 Accept 和 User-Agent 头，如果配置了 Token 则附加 Authorization 头。
     *
     * @returns HTTP 请求头对象
     */
    private async buildHeaders(): Promise<Record<string, string>> {
        const token = await this.config.getValueDefault('github.token', '');
        const h: Record<string, string> = { Accept: 'application/vnd.github.v3+json', 'User-Agent': 'GithubStars-Search' };
        if (token) h['Authorization'] = `Bearer ${token}`;
        return h;
    }

    /**
     * 搜索 GitHub 仓库
     *
     * 支持关键词和语言筛选，调用 GitHub Search API 返回仓库列表。
     * 搜索失败时返回 success: false 和错误信息，不抛出异常。
     *
     * @param keyword 搜索关键词
     * @param language 编程语言
     * @param sort 排序字段，默认 stars
     * @param page 页码，从 1 开始
     * @param perPage 每页数量，默认 20
     * @returns 搜索结果对象，包含 success、total、repos、page、perPage
     */
    async searchRepos(keyword: string, language: string, sort = 'stars', page = 1, perPage = 20) {
        this.logger.log('搜索仓库: keyword=' + keyword + ', language=' + language + ', page=' + page);
        try {
            let q = keyword || '';
            if (language) q += ` language:${language}`;
            if (!q.trim()) q = 'stars:>1';
            const params = new URLSearchParams({ q, sort: sort || 'stars', page: String(page), per_page: String(perPage) });
            const res = await fetch(`${GITHUB_API}/search/repositories?${params}`, { headers: await this.buildHeaders() });
            if (res.status === 200) {
                const data = await res.json();
                const repos = (data.items || []).map((item: any) => ({
                    id: item.id,
                    fullName: item.full_name || '',
                    description: item.description || '',
                    language: item.language || '',
                    starsCount: item.stargazers_count || 0,
                    forksCount: item.forks_count || 0,
                    htmlUrl: item.html_url || '',
                    pushedAt: item.pushed_at || '',
                    createdAt: item.created_at || '',
                    ownerName: item.owner?.login || '',
                    ownerAvatarUrl: item.owner?.avatar_url || '',
                    topics: Array.isArray(item.topics) ? item.topics : [],
                }));
                return { success: true, total: data.total_count || 0, repos, page, perPage };
            }
            if (res.status === 403) throw new Error('GitHub API rate limited');
            throw new Error(`GitHub API error: ${res.status}`);
        } catch (e) {
            const msg = e instanceof Error ? e.message : String(e);
            this.logger.error('搜索仓库失败: ' + msg);
            if (msg.includes('rate limited')) return { success: false, total: 0, repos: [], page, perPage, message: 'GitHub API 限流' };
            return { success: false, total: 0, repos: [], page, perPage, message: `搜索失败: ${msg}` };
        }
    }

    /**
     * Star 一个 GitHub 仓库
     *
     * 向 GitHub API 发送 PUT 请求标记星标。
     *
     * @param owner 仓库所有者
     * @param repo 仓库名
     * @returns 成功返回 true
     */
    async starRepo(owner: string, repo: string): Promise<boolean> {
        this.logger.log('Star 仓库: ' + owner + '/' + repo);
        const token = await this.config.getValueDefault('github.token', '');
        if (!token) {
            this.logger.error('Star 仓库失败: 未配置 github.token');
            return false;
        }
        try {
            const res = await fetch(`${GITHUB_API}/user/starred/${owner}/${repo}`, {
                method: 'PUT',
                headers: { ...(await this.buildHeaders()), 'Content-Length': '0' },
            });
            if (res.status === 204 || res.status === 304) return true;
            this.logger.error(`Star 仓库失败: ${owner}/${repo}, GitHub 响应 status=${res.status}`);
            return false;
        } catch (e) {
            this.logger.error(`Star 仓库异常: ${owner}/${repo}`, e instanceof Error ? e : undefined);
            return false;
        }
    }

    /**
     * 取消 Star 一个 GitHub 仓库
     *
     * 向 GitHub API 发送 DELETE 请求取消星标。
     *
     * @param owner 仓库所有者
     * @param repo 仓库名
     * @returns 成功返回 true
     */
    async unstarRepo(owner: string, repo: string): Promise<boolean> {
        this.logger.log('取消 Star 仓库: ' + owner + '/' + repo);
        try {
            const res = await fetch(`${GITHUB_API}/user/starred/${owner}/${repo}`, {
                method: 'DELETE',
                headers: await this.buildHeaders(),
            });
            return res.status === 204;
        } catch (e) {
            this.logger.error(`取消 Star 仓库失败: ${owner}/${repo}`, e instanceof Error ? e : undefined);
            return false;
        }
    }

    /**
     * 检查是否已 Star 某个仓库
     *
     * 向 GitHub API 发送 GET 请求检查星标状态。
     *
     * @param owner 仓库所有者
     * @param repo 仓库名
     * @returns 已 Star 返回 true
     */
    async checkStarred(owner: string, repo: string): Promise<boolean> {
        try {
            const res = await fetch(`${GITHUB_API}/user/starred/${owner}/${repo}`, { headers: await this.buildHeaders() });
            return res.status === 204;
        } catch (e) {
            this.logger.error(`检查 Star 状态失败: ${owner}/${repo}`, e instanceof Error ? e : undefined);
            return false;
        }
    }
}
